### PR TITLE
Add Strava applet

### PR DESCRIPTION
**Pull request by rgkimball**
Hello!

The applet in this PR implements a simple display for summary stats from [Strava](https://strava.com) using their API. You log in using OAuth and can display data from your runs, rides or swims either all-time or your YTD stats only. Currently the application is under my user, but happy to transfer this if the Tidbyt team feels that would be easier to maintain! Only the client id and secret would need to be replaced.

IMG_8355

Here are a few more images from the Pixlet environment showing other options and the relating config:

- Runs: https://i.imgur.com/h5dX9Ox.png
- Swims: https://i.imgur.com/LSlfvgI.png (apologies to swimmers for the lack of personal data here!)
- YTD Rides: https://i.imgur.com/3wAlDkP.png

Happy to hear/incorporate any feedback you may have.

### CODE REVIEW

**Comment by rgkimball** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by rohansingh** [comment]
@rgkimball If you don't mind, I'm just going to push a commit to fix the lint issues and get this merged.

**Comment by rgkimball** [comment]
> @rgkimball If you don't mind, I'm just going to push a commit to fix the lint issues and get this merged.

No objections, push away!

File: apps/strava/strava.star [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+             (res.status_code, res.body()))
+
+    token_params = res.json()
+    refresh_token = token_params["refresh_token"]
+    access_token = token_params["access_token"]
+    athlete = int(float(token_params['athlete']['id']))
+
+    cache.set(GLOBAL_CACHE_PREFIX + 'athlete_id', str(athlete), ttl_seconds=CACHE_TTL)
+    cache.set(GLOBAL_CACHE_PREFIX + "access_token", access_token, ttl_seconds = int(token_params["expires_in"] - 30))
+    cache.set(GLOBAL_CACHE_PREFIX + "refresh_token", refresh_token, ttl_seconds = int(token_params["expires_in"] - 30))
```
Comment (zephyern) [inline apps/strava/strava.star:300]: It's my understanding that the cache is global across all users. Are you sure you can store these tokens in the global cache? Wouldn't that mean that this app would only work for the first user who ever authenticated?

Comment (rgkimball) [reply apps/strava/strava.star:300]: Is that true? I can include the athlete ID in the key name if that's the case but would appreciate if someone could confirm that the cache is not local to the device.

Comment (zephyern) [reply apps/strava/strava.star:300]: This was my understanding after reading the docs here (https://github.com/tidbyt/pixlet/blob/main/docs/authoring_apps.md):

> Make sure to create cache keys that are unique for the type of information you are caching. Each app has its own cache, but that cache is shared among every user. Two installations of the same app by two different users will share the same cache.

Comment (rgkimball) [reply apps/strava/strava.star:300]: Ok, is there an "instance ID" or something that I can include so people can have multiple Strava views on one device?

Comment (zephyern) [reply apps/strava/strava.star:300]: I just asked the same question in my pull request. :) I was checking out your code to see how you were doing OAuth, as I'm not quite sure how to test it.

It's really clever how you put the token exchange in the render function. I could try that to test that part of my code.

Can I ask, how did you test the initial OAuth request? Did you just have it redirect the response to a server you own? Thanks!

Comment (rgkimball) [reply apps/strava/strava.star:300]: Thanks! I was able to test the oauth exchange using pixlet serve, you can test your config that way as well. 

The one thing I couldn't get working locally was the secret encryption, I assume only apps on device have access to the private key?

Comment (zephyern) [reply apps/strava/strava.star:300]: Oh, thanks for the advice! Let me try `pixlet serve` again.

Re: Secrets
Yep! (Same doc as above.)

> When you run pixlet locally, secret.decrypt will always return None. When your app runs in the Tidbyt cloud, secret.decrypt will return the string that you passed to pixlet encrypt.

Comment (rgkimball) [reply apps/strava/strava.star:300]: Got it, thanks!!

**Comment by rohansingh** [comment]
@rgkimball @zephyern After reading your discussion, I ended up making several edits here to get the auth working for multiple users.

First off, just to address this:

> It's my understanding that the cache is global across all users. Are you sure you can store these user-specific tokens in the global cache? Wouldn't that mean that this app would only work for the first user who ever authenticated?

This is totally correct. The cache is scoped per app, but it's global to all users of that app. Here's how we handle this for OAuth apps and store user-specific stuff:

1. Return the refresh token from `oauth_handler`. This token is permanently saved for that particular user.
2. Cache other stuff using the refresh token as part of the cache key.

Here's an example of how it's done in the Netatmo app:
https://github.com/tidbyt/community/blob/0c65a014ea92a86c920bf2f1a4dfcf60571be293/apps/netatmo/netatmo.star#L132

I modified the code for this app to do the same thing. The refresh token is returned from `oauth_handler`, and stuff like the access token and athlete ID are cached using the refresh token as the cache key.

---

The big reason that I _had_ to make this change is because `oauth_handler` needs to return the refresh token if it's available. This is because both the authorization code and the access token will expire after a while. So @zephyern:

> It's really clever how you put the token exchange in the render function. I could try that to test that part of my code.

That actually won't work. It will actually seem like it works locally for a bit, but then the token exchange will start failing after a while because the authorization code is only valid for a few minutes.

---

Finally, I changed the flow to return the preview stats if there's nobody logged in. We need to do this to have a real preview show up in the Tidbyt phone app, so that user's know what the app looks like.

**Comment by rohansingh** [comment]
Rebasing to fix tests.

**Pull request merged**